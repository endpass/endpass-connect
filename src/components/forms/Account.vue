<template>
  <form @submit.prevent="emitSubmit">
    <form-field v-if="message">
      <message :success="true" data-test="success-message">{{
        message
      }}</message>
    </form-field>
    <form-field v-if="error">
      <message :error="true" data-test="error-message">{{ error }}</message>
    </form-field>
    <form-field label="Active account address">
      <v-select
        v-model="formData.activeAccount"
        :items="accounts"
        name="activeAccount"
      />
    </form-field>
    <form-field label="Active network">
      <v-select
        v-model="formData.activeNet"
        :items="networks"
        name="activeNetwork"
      />
    </form-field>
    <form-field>
      <v-button
        :submit="true"
        :disabled="loading"
        type="primary"
        data-test="submit-button"
        fluid
        >{{ primaryButtonLabel }}</v-button
      >
    </form-field>
    <v-faucet-button
      v-if="isRopsten"
      :address="formData.activeAccount"
      class-name="button primary fluid"
      @before-request="emitDonateRequest"
      @donate="emitDonateSuccess"
      @donate-error="emitDonateError"
    >
      Request 1 ETH from faucet
    </v-faucet-button>
    <form-controls>
      <v-button
        :disabled="loading"
        type="danger"
        data-test="logout-button"
        @click="emitLogout"
        >Logout</v-button
      >
      <v-button
        :disabled="!closable || loading"
        data-test="cancel-button"
        @click="emitCancel"
        >Close</v-button
      >
    </form-controls>
  </form>
</template>

<script>
import { VFaucetButton } from '@endpass/faucet';
import { DEFAULT_NETWORKS } from '@/constants';
import VButton from '../VButton.vue';
import VSelect from '../VSelect.vue';
import Message from '../Message.vue';
import FormField from '../FormField.vue';
import FormControls from '../FormControls.vue';

export default {
  name: 'AccountForm',

  props: {
    closable: {
      type: Boolean,
      default: true,
    },

    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    message: {
      type: String,
      default: null,
    },

    accounts: {
      type: Array,
      default: () => [],
    },

    networks: {
      type: Array,
      default: () => [],
    },

    formData: {
      type: Object,
      required: true,
    },
  },

  computed: {
    primaryButtonLabel() {
      return !this.loading ? 'Update account' : 'Loading...';
    },

    isRopsten() {
      const ropstenNet = Object.values(DEFAULT_NETWORKS).find(
        net => net.networkType === 'ropsten',
      );

      /* eslint-disable-next-line */
      return this.formData.activeNet == ropstenNet.id;
    },
  },

  methods: {
    emitSubmit() {
      if (!this.loading) {
        this.$emit('submit');
      }
    },

    emitLogout() {
      if (!this.loading) {
        this.$emit('logout');
      }
    },

    emitCancel() {
      if (this.closable) {
        this.$emit('cancel');
      }
    },

    emitDonateRequest() {
      this.$emit('donate-request');
    },

    emitDonateSuccess(e) {
      this.$emit('donate-success', e);
    },

    emitDonateError(e) {
      if (e.message.includes('403')) {
        this.$emit(
          'donate-error',
          'Something went wrong. Try request dontaion later!',
        );
      }
    },
  },

  components: {
    VFaucetButton,
    VButton,
    VSelect,
    Message,
    FormField,
    FormControls,
  },
};
</script>
