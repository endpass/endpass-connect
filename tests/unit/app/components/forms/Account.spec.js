import { shallowMount, mount } from '@vue/test-utils';
import Account from '@/components/forms/Account.vue';

describe('Account', () => {
  const componentOptions = {
    propsData: {
      formData: {
        activeAccount: '0x0',
        activeNet: 1,
      },
    },
  };
  let wrapper;

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallowMount(Account, componentOptions);
    });

    it('should correctly render Account component', () => {
      expect(wrapper.name()).toBe('AccountForm');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render faucet button only if activeNet is ropsten', () => {
      wrapper.setProps({
        formData: {
          activeNet: 3,
        },
      });

      expect(wrapper.find('v-faucet-button-stub').exists()).toBe(true);
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      wrapper = mount(Account, componentOptions);
    });

    describe('submit feature', () => {
      it('should emit submit on click submit button by default', () => {
        wrapper.find('form').vm.$emit('submit');

        expect(wrapper.emitted().submit).toBeTruthy();
      });
    });

    describe('cancel feature', () => {
      beforeEach(() => {
        wrapper = mount(Account, componentOptions);
      });

      it('should emit close on click cancel button by default', () => {
        wrapper.find('[data-test=cancel-button]').vm.$emit('click');

        expect(wrapper.emitted().cancel).toBeTruthy();
      });
    });
  });
});
