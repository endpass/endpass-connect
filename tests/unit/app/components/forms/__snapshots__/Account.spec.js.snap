// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account render should correctly render Account component 1`] = `
<form>
  <!---->
  <!---->
  <form-field-stub label="Active account address">
    <v-select-stub items="" name="activeAccount" value="0x0"></v-select-stub>
  </form-field-stub>
  <form-field-stub label="Active network">
    <v-select-stub items="" name="activeNetwork" value="1"></v-select-stub>
  </form-field-stub>
  <form-field-stub>
    <v-button-stub type="primary" submit="true" fluid="true" data-test="submit-button">Update account</v-button-stub>
  </form-field-stub>
  <!---->
  <form-controls-stub>
    <v-button-stub type="danger" data-test="logout-button">Logout</v-button-stub>
    <v-button-stub data-test="cancel-button">Close</v-button-stub>
  </form-controls-stub>
</form>
`;
